wsd/COOLWSD.cpp

```diff
@@ -3274,6 +3277,103 @@ class ClientRequestDispatcher final : public SimpleSocketHandler
         LOG_INF("Sent capabilities.json successfully.");
     }
 
+    void handleCheckRequest(const Poco::Net::HTTPRequest& request,
+                                   const std::shared_ptr<StreamSocket>& socket,
+                                   Poco::MemoryInputStream& message)
+    {
+        assert(socket && "Must have a valid socket");
+
+        LOG_DBG("Check request: " << request.getURI());
+
+        bool validWopiConfig = false;
+        std::string connectMessage;
+        bool validConnectBack = false;
+
+        ConvertToPartHandler handler;
+        HTMLForm form(request, message, handler);
+
+        std::string url = (form.has("url") ? form.get("url") : "");
+        Poco::URI remoteServerURI(url);
+
+        // Validate allowed WOPI host
+        std::string addressToCheck = remoteServerURI.getHost();
+        LOG_DBG("Checking remote host: " << addressToCheck);
+        try
+            {
+                if (!StorageBase::allowedWopiHost(addressToCheck) && !net::isLocalhost(addressToCheck))
+                {
+                    // FIXME: does not automatically detect that 127.0.0.1 matches localhost in coolwsd.xml
+                    const auto hostAddresses(Poco::Net::DNS::resolve(addressToCheck));
+                    for (auto &address : hostAddresses.addresses())
+                    {
+                        LOG_DBG("Checking resolved remote host: " << address.toString());
+                        if (StorageBase::allowedWopiHost(address.toString()))
+                        {
+                            validWopiConfig = true;
+                            break;
+                        }
+                        LOG_DBG("Invalid resolved remote host: " << address.toString());
+                    }
+                } else {
+                    validWopiConfig = true;
+                }
+
+            }
+            catch (const Poco::Exception& exc)
+            {
+                LOG_ERR("Poco::Net::DNS::resolve(\"" << addressToCheck << "\") failed: " << exc.displayText());
+                // We can't find out the hostname, and it already failed the IP check
+                validWopiConfig = false;
+            }
+
+        // Validate connecting back to WOPI host
+        try
+            {
+                std::shared_ptr<http::Session> httpSession(
+                    StorageBase::getHttpSession(remoteServerURI));
+                http::Request wopiRequest(remoteServerURI.getPathAndQuery());
+
+                const std::shared_ptr<const http::Response> httpResponse =
+                    httpSession->syncRequest(wopiRequest);
```

Comment:
Sync is synchronous - this will block the server and the main thread until that request completes. We have to do async requests.